File: services/api/src/aec_api/deal_memory.py
"""R35-DEAL-MEMORY — the operator's own closed projects as a comp set, and the comps it won't invent.

When somebody types an assumption into an underwriting, the most useful thing on the screen is what
this firm's *own* past projects actually did. That layer cannot be bought: it is made of the
operator's history, which is why it is the least commoditised part of the stack.

It is also the easiest place in the product to manufacture a flattering number, and the failure has
a specific shape:

**"Realised" must come from a different record than "assumed", or the comparison is circular.** The
roadmap asks for *exit cap achieved vs assumed* and *actual lease-up months*. Neither is recorded
anywhere in this system — there is no disposition record and no lease-up actual. The tempting move
is to read them off the latest scenario, but the latest scenario is an **assumption**. Comparing it
to the assumption being entered would report near-perfect underwriting accuracy forever, and that
number would then be used to justify the next deal. So those metrics are reported as
`no_recorded_source`, naming what would have to be captured, and are never computed.

What IS recorded, and is therefore reported:

* **cost variance** — `cost.summary` carries budget *and* actual, from commitments and invoices;
* **cost per SF** — the same actual over the model's own GFA;
* **schedule variance** — `schedule_activity` carries `finish` *and* `actual_finish`.

Each is a *measurement* sitting in a different record from the plan it is compared against. Two
further refusals, both shared with `deal_funnel`:

* **a sample floor.** Under `MIN_SAMPLES` closed projects a distribution is noise wearing a
  percentage sign, so no distribution is emitted — the count is, so the gap is visible;
* **a project with no actuals is not a project with zero variance.** It is excluded and counted;
  averaging it in as 0% would drag every distribution toward "we were exactly right".
"""
from __future__ import annotations

from datetime import date, datetime
from typing import Any

#: Closed projects a metric needs before its distribution is reported at all.
MIN_SAMPLES = 5

STATUS_OK = "ok"
STATUS_INSUFFICIENT = "insufficient_history"
STATUS_NO_SOURCE = "no_recorded_source"

#: Requested by R35-DEAL-MEMORY but with NO realised record anywhere in the system. Named rather
#: than silently omitted: a metric that is absent looks like a metric that was fine.
UNSOURCED: dict[str, str] = {
    "exit_cap_achieved": (
        "no disposition record exists — nothing captures the cap rate a sale actually achieved. "
        "The only exit cap in the system is the ASSUMED one on a scenario, so comparing it to the "
        "assumption being entered would compare a number to itself and report perfect accuracy "
        "forever. Needs a disposition record (sale date, price, trailing NOI) before it can be real."),
    "lease_up_months": (
        "no lease-up actual is recorded — there is no first-unit-leased or stabilisation date to "
        "measure from. The absorption figures in the system are forecasts, not observations."),
}


def _num(v: Any) -> float | None:
    if v is None or isinstance(v, bool) or v == "":
        return None
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    return None if f != f or f in (float("inf"), float("-inf")) else f


def _d(v: Any) -> date | None:
    if isinstance(v, datetime):
        return v.date()
    if isinstance(v, date):
        return v
    if not v:
        return None
    try:
        return date.fromisoformat(str(v)[:10])
    except ValueError:
        return None


def distribution(values: list[float], min_samples: int = MIN_SAMPLES) -> dict[str, Any]:
    """p25 / median / p75 over the observations, or a refusal naming how many more are needed."""
    n = len(values)
    if n < min_samples:
        return {"status": STATUS_INSUFFICIENT, "count": n, "needed": min_samples - n,
                "median": None, "p25": None, "p75": None,
                "note": (f"{n} closed project(s) carry this measurement; {min_samples - n} more "
                         "needed before a distribution means anything. No default is substituted.")}
    s = sorted(values)

    def q(p: float) -> float:
        if len(s) == 1:
            return s[0]
        i = p * (len(s) - 1)
        lo, hi = int(i), min(int(i) + 1, len(s) - 1)
        return s[lo] + (s[hi] - s[lo]) * (i - lo)

    return {"status": STATUS_OK, "count": n, "needed": 0,
            "median": round(q(0.5), 4), "p25": round(q(0.25), 4), "p75": round(q(0.75), 4),
            "min": round(s[0], 4), "max": round(s[-1], 4)}


def _schedule_variance_days(db, pid: str) -> float | None:
    """Days between the planned and actual finish of the project's LAST activity to finish.

    The project's finish is the latest finish among its activities, so the variance that matters is
    measured on the activity that actually ended last — not an average across activities, which
    would let an early trade's overrun cancel a late trade's recovery.
    """
    from . import modules as me
    try:
        rows = me.list_records(db, "schedule_activity", pid, limit=1_000_000)
    except Exception:                           # noqa: BLE001 — see below
        # `list_records` raises HTTPException(404) when the module is not registered (a trimmed
        # deployment, or a registry not yet loaded). A portfolio roll-up must not be able to die
        # because one module is absent: the honest answer for this project is "no measurement",
        # which is already how a project with no schedule is handled two lines down.
        return None
    planned_last = actual_last = None
    for r in rows:
        d = r.get("data") or {}
        p, a = _d(d.get("finish")), _d(d.get("actual_finish"))
        if p and (planned_last is None or p > planned_last):
            planned_last = p
        if a and (actual_last is None or a > actual_last):
            actual_last = a
    if planned_last is None or actual_last is None:
        return None
    return float((actual_last - planned_last).days)


def comps(db, project_ids: set[str] | None = None, min_samples: int = MIN_SAMPLES) -> dict[str, Any]:
    """Realised outcomes across the caller's own closed projects.

    `project_ids=None` means no restriction (RBAC off / admin); otherwise scoped exactly as
    `fca.portfolio` and `benchmarking` are, so one tenant's history never becomes another's comps.
    """
    from . import cost as cost_engine
    from . import energy
    from .models import Project

    q = db.query(Project)
    if project_ids is not None:
        q = q.filter(Project.id.in_(project_ids))

    cost_var: list[float] = []
    cost_sf: list[float] = []
    sched_var: list[float] = []
    rows: list[dict[str, Any]] = []
    excluded: list[dict[str, Any]] = []

    for p in q.all():
        try:
            cs = cost_engine.summary(db, p.id)
        except Exception:                       # noqa: BLE001 — a project with no cost data still lists
            cs = {}
        budget = _num(cs.get("budget"))
        actual = _num(cs.get("actual"))
        gfa = None
        try:
            gfa = _num(energy.project_gfa_sf(db, p.id))
        except Exception:                       # noqa: BLE001
            gfa = None
        sv = _schedule_variance_days(db, p.id)

        rec: dict[str, Any] = {"id": p.id, "name": p.name,
                               "vintage": p.created_at.year if p.created_at else None,
                               "budget": budget, "actual": actual, "gfa_sf": gfa,
                               "cost_variance_pct": None, "cost_per_sf": None,
                               "schedule_variance_days": sv}

        # A project with no ACTUAL spend has not been built yet. Counting it as 0% variance would
        # drag every distribution toward "we were exactly right" — the most flattering artefact
        # available and the one nobody would question.
        if budget and actual:
            rec["cost_variance_pct"] = round((actual - budget) / budget * 100.0, 2)
            cost_var.append(rec["cost_variance_pct"])
            if gfa:
                rec["cost_per_sf"] = round(actual / gfa, 2)
                cost_sf.append(rec["cost_per_sf"])
        else:
            excluded.append({"id": p.id, "name": p.name, "reason": "no_actual_cost",
                             "note": "no actual spend recorded; excluded rather than counted as "
                                     "zero variance"})
        if sv is not None:
            sched_var.append(sv)
        rows.append(rec)

    unsourced = {k: {"status": STATUS_NO_SOURCE, "note": v} for k, v in UNSOURCED.items()}
    return {
        "project_count": len(rows),
        "metrics": {
            "cost_variance_pct": distribution(cost_var, min_samples),
            "cost_per_sf": distribution(cost_sf, min_samples),
            "schedule_variance_days": distribution(sched_var, min_samples),
            **unsourced,
        },
        "projects": rows,
        "excluded": excluded,
        "excluded_count": len(excluded),
        "min_samples": min_samples,
        "note": ("every reported metric is a MEASUREMENT held in a different record from the plan it "
                 "is compared against. Metrics marked no_recorded_source are not computed at all — "
                 "deriving them from a scenario would compare an assumption to itself."),
    }


def beside(metric: str, value: Any, memory: dict[str, Any]) -> dict[str, Any]:
    """The firm's own realised distribution for `metric`, next to the value being entered.

    This is the shape the underwriting screen wants: not a verdict, a comparison. It reports where
    the entered value sits against the operator's own history and refuses to grade it when there is
    no history to grade against.
    """
    m = (memory.get("metrics") or {}).get(metric)
    if m is None:
        return {"metric": metric, "status": "unknown_metric", "entered": value,
                "known": sorted((memory.get("metrics") or {}).keys())}
    if m.get("status") != STATUS_OK:
        return {"metric": metric, "status": m.get("status"), "entered": value,
                "comparison": None, "note": m.get("note")}
    v = _num(value)
    if v is None:
        return {"metric": metric, "status": "not_a_number", "entered": value, "comparison": None}
    lo, hi, med = m["p25"], m["p75"], m["median"]
    where = "below_p25" if v < lo else "above_p75" if v > hi else "within_iqr"
    return {"metric": metric, "status": STATUS_OK, "entered": v, "median": med,
            "p25": lo, "p75": hi, "count": m["count"], "position": where,
            "note": (f"this firm's own {m['count']} closed project(s) landed between {lo} and {hi} "
                     f"(median {med}). This is a comparison, not a verdict — the history says where "
                     "the number sits, not whether it is right.")}

File: apps/web/src/api/dealMemory.ts
/** R35-DEAL-MEMORY: this firm's own realised $/SF, beside the hard cost being underwritten.
 *
 *  A new file rather than methods in `client.ts`, per the extraction rule in `test_file_sizes.py`:
 *  a *field* folds onto an existing line, a *feature* moves to a domain module. **Deliberately NOT
 *  labelled SCALE-SEAM ㉘** — a seam slice EXTRACTS an existing group and shrinks `client.ts`, and
 *  this adds a feature that was never in it. The ratchet said so before I did: the one import line
 *  took the file 2,837 → 2,838 and failed, which is a slice claim failing its own definition.
 *
 *  **`deal_memory.beside()` — the function whose own docstring says it is "the shape the underwriting
 *  screen wants" — had no caller anywhere**, server or client, while `deal_memory.comps` had been
 *  routed since the engine shipped. So `test_reachable` passed: the MODULE is imported, by the
 *  portfolio route. *A module can be reachable and its whole reason for existing still be
 *  unreachable* — `read_p6xml_all` was the same shape one ring over.
 *
 *  **Both methods are here, and the second one is here because two gates argued about it.**
 *  `portfolioDealMemory` was written, then removed when `clientCallers.test.ts` failed on it — no
 *  screen, so a typed method for it would exist only to satisfy a gate. Then
 *  `test_route_reachability` failed the other way: adding `/projects/{pid}/deal-memory/beside` put
 *  the substring `deal-memory` into the web source, so its frozen entry for `/portfolio/deal-memory`
 *  read as "quietly became called" when nothing had called it.
 *
 *  Neither gate was wrong and neither could be satisfied by wording. What settled it was re-reading
 *  the roadmap item, which asks for realised outcomes *"(exit cap achieved vs assumed, actual
 *  lease-up months, **cost/SF by vintage**)"* — and `comps` returns `vintage` per project. The comp
 *  list was part of the item all along; wiring only the summary comparison was the half-build, and
 *  the gate collision is what made that visible.
 *
 *  A mixin, so every call site resolves unchanged. `api/surface.test.ts` is what proves it.
 */
import { HttpCore } from "./httpCore";

type Ctor<T> = new (...args: any[]) => T;

/** `beside()`'s answer: where an entered number sits against the firm's own history. */
export interface DealMemoryBeside {
  metric: string;
  /** `no_gfa` is the route's own, and is NOT a `beside()` status — see the method's comment. */
  status: "ok" | "insufficient_history" | "no_recorded_source" | "unknown_metric" | "not_a_number"
    | "no_gfa";
  entered: number | null;
  median?: number; p25?: number; p75?: number; count?: number;
  position?: "below_p25" | "within_iqr" | "above_p75";
  comparison?: null;
  note?: string;
  gfa_sf?: number;
  hard_cost?: number;
}

/** One metric's realised distribution across the caller's own closed projects. */
export interface DealMemoryMetric {
  status: "ok" | "insufficient_history" | "no_recorded_source";
  count?: number;
  median?: number; p25?: number; p75?: number; min?: number; max?: number;
  note?: string;
}

/** One closed project as it contributes to the comp set. */
export interface DealMemoryProject {
  id: string; name: string; vintage: number | null;
  budget: number | null; actual: number | null; gfa_sf: number | null;
  cost_variance_pct: number | null; cost_per_sf: number | null;
  schedule_variance_days: number | null;
}

export function withDealMemory<TBase extends Ctor<HttpCore>>(Base: TBase) {
  return class DealMemory extends Base {
  /** The closed projects behind the comparison — the "by vintage" half of the roadmap item.
   *
   *  `exit_cap_achieved` and `lease_up_months` come back as `no_recorded_source` and are never
   *  computed: nothing records a realised exit cap or a lease-up actual, and reading either off the
   *  latest scenario would compare an assumption to itself and report perfect underwriting accuracy
   *  forever. Render the refusal — it is the honest half of the payload.
   *
   *  `excluded` is not noise either: a project with no actual spend is excluded rather than counted
   *  as zero variance, and showing the count is what stops a small comp set looking like a complete
   *  one. */
  portfolioDealMemory() {
    return this.json<{
      project_count: number;
      metrics: Record<string, DealMemoryMetric>;
      projects: DealMemoryProject[];
      excluded: { id: string; name: string; reason: string; note: string }[];
      excluded_count: number; min_samples: number; note: string;
    }>("/portfolio/deal-memory");
  }

  /** This firm's own realised $/SF, beside the hard cost being underwritten.
   *
   *  `hard_cost` is dollars, not $/SF: the server divides by the project's GFA because
   *  `energy.project_gfa_sf` is the one definition of GFA and deriving it here would make a second.
   *  A project whose properties index is not loaded answers `no_gfa` — a real and common outcome,
   *  and deliberately distinct from `insufficient_history`: one means "we cannot express your number
   *  in this metric's units", the other means "we have no history to compare it against", and they
   *  send a reader looking in different places. */
  dealMemoryBeside(pid: string, hardCost: number) {
    return this.json<DealMemoryBeside>(
      `/projects/${pid}/deal-memory/beside?hard_cost=${encodeURIComponent(String(hardCost))}`);
  }
  };
}
